Removes tab cycling between TGUI say channels in asay

Kubisopplay opened the pull request:
## About The Pull Request
Removes the capability to cycle input channels by pressing tab when the channel open is asay

## Why It's Good For The Game
Lowers chance to leak asay into ooc. I heard asay is supposed to be secret, so this sounds like an improvement.

## Changelog

:cl:
admin: Lowered the chance of accidentally leaking asay
/:cl:

Kubisopplay committed:
Removes tab cycling from tgui asay

Kubisopplay committed:
Stops you from cycling INTO asay too.

Mothblocks commented on tgui/packages/tgui-say/handlers/incrementChannel.tsx:
```diff
@@ -12,8 +12,15 @@ export const handleIncrementChannel = function (this: Modal) {
+  }
   this.fields.radioPrefix = '';
-  if (channel === CHANNELS.length - 1) {
+  if (channel === CHANNELS.length - 2) {
```
can we comment what this number is

Kubisopplay replied:
I kinda wanted to do this with minimal changes to the code. 1 was just cycling back to the start of the array. 2 is cycling back from the second to last element, so ignoring asay. I don't really know how to comment it better...

Mothblocks replied:
that seems *really* fragile. if we add a new channel to the end then we're going to be skipping over that instead. i dont like this

Kubisopplay replied:
I can write an explanation why, and add a reminder in a comment in the definition file to keep the asay as the last one in array, is that okay?

Kubisopplay replied:
To be fair the entire proc should be stripped from magic numbers, since OOC is treated the same way. I might do that tomorrow, but for now I wanted to block the thing that multiple admins complained about already.

Mothblocks replied:
> I can write an explanation why, and add a reminder in a comment in the definition file to keep the asay as the last one in array, is that okay?

comments aren't a substitute for good code :(

Kubisopplay replied:
Uhh, I changed it, can you look at it now?

Kubisopplay committed:
More futureproof solution

I think? (The process can't be called thinking, more of randomly trying things)

Mothblocks commented on tgui/packages/tgui-say/handlers/incrementChannel.tsx:
```diff
@@ -24,9 +36,19 @@ export const handleIncrementChannel = function (this: Modal) {
+      channel === ADMIN_CHANNEL - 1 &&
+      CHANNELS.length - 1 !== ADMIN_CHANNEL
+    ) {
+      this.setState({
+        buttonContent: CHANNELS[channel + 2],
+        channel: channel + 2,
+      });
```
Instead of this, you should have a set of names to blacklist. Then, add 1 to channel until you get something not on the blacklist, or wrap around to the beginning (where you will continue to check until you get a valid result)

Kubisopplay replied:
Can I add the same behavior for OOC then? so it can only cycle within the IC buttons? If I am going to rework it into a blacklist anyway.

Kubisopplay committed:
Hate linters HATE LINTERS

Kubisopplay committed:
General refactor?

Im going to sleep

Kubisopplay commented:
Alright, refactored it into something that is hopefully modular

Kubisopplay commented:
Hello? Could I get a look at this?

Mothblocks commented:
its been a day relax

Mothblocks commented on tgui/packages/tgui-say/handlers/incrementChannel.tsx:
```diff
@@ -1,32 +1,49 @@
 import { Modal } from '../types';
+// Insert the names of channels you want to not cycle on tab here
+const BLACKLIST = ['Admin'];
+const BLACKLISTED_CHANNELS = [] as number[];
+CHANNELS.forEach((channel, index) => {
+  if (BLACKLIST.includes(channel)) {
+    BLACKLISTED_CHANNELS.push(index);
+  }
+});
```
Change this whole thing into a map, e.g. `BLACKLIST.map((blacklistedChannel) => CHANNELS.findIndex(blacklistedChannel)` or whatevs, but not this

Also you should change to BLACKLISTED_CHANNEL_INDICES or similar, because right now this is just a synonym

Kubisopplay replied:
Alright, done

Kubisopplay committed:
Requested changes

Kubisopplay committed:
Im stupdi

Mothblocks commented on tgui/packages/tgui-say/handlers/incrementChannel.tsx:
```diff
@@ -1,32 +1,48 @@
 import { Modal } from '../types';
+// Insert the names of channels you want to not cycle on tab here
+const BLACKLIST = ['Admin'];
+const BLACKLISTED_CHANNEL_INDICES = CHANNELS.map((channel, index) => {
```
Nit: We have `filterMap` for this in collections.ts

Kubisopplay replied:
Filtermap requires having not readonly

Mothblocks replied:
It shouldn't if you do something like:

`filterMap(CHANNELS, (channel: readonly string, index)` or something. What error were you getting with what code?

Kubisopplay replied:
Linter says "Argument of type 'readonly ["Say", "Radio", "Me", "OOC", "Admin"]' is not assignable to parameter of type 'unknown[]'.
  The type 'readonly ["Say", "Radio", "Me", "OOC", "Admin"]' is 'readonly' and cannot be assigned to the mutable type 'unknown[]'.". The code compiles, but doesnt work. Trying to get it working only caused more problems, since afaik filtermap doesnt use two argument lambda. I think that simplicity of using filter and map separately is good enough here.

Mothblocks replied:
Dang, oh well

Mothblocks commented on tgui/packages/tgui-say/handlers/incrementChannel.tsx:
```diff
@@ -1,32 +1,48 @@
-      buttonContent: CHANNELS[0],
-      channel: 0,
+      buttonContent: CHANNELS[channel],
+      channel: channel,
```
```suggestion
      channel,
```

Mothblocks commented on tgui/packages/tgui-say/handlers/incrementChannel.tsx:
```diff
@@ -1,32 +1,48 @@
 import { CHANNELS } from '../constants';
 import { Modal } from '../types';
+// Insert the names of channels you want to not cycle on tab here
```
```suggestion

// Insert the names of channels you want to not cycle on tab here
```

Mothblocks commented on tgui/packages/tgui-say/handlers/incrementChannel.tsx:
```diff
@@ -1,32 +1,48 @@
   }
+  this.setState({
+    buttonContent: CHANNELS[channel],
+    channel: channel,
```
```suggestion
    channel,
```

Kubisopplay committed:
Update tgui/packages/tgui-say/handlers/incrementChannel.tsx

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Kubisopplay committed:
Update tgui/packages/tgui-say/handlers/incrementChannel.tsx

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Kubisopplay committed:
Update tgui/packages/tgui-say/handlers/incrementChannel.tsx

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>
